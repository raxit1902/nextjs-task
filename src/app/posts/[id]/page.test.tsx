import { render, screen } from "@testing-library/react";
import PostPageWrapper from "./page"; // Import the page.tsx file
import "@testing-library/jest-dom";

global.fetch = jest.fn();

describe("PostPageWrapper Component", () => {
    beforeEach(() => {
        jest.resetAllMocks(); // Reset all mocks before each test
    });

    test("renders post details correctly when API call is successful", async () => {
        (global.fetch as jest.Mock).mockResolvedValueOnce({
            ok: true,
            json: async () => ({
                id: "1",
                title: "Sample Post",
                body: "This is a sample post body.",
            }),
        });

        const params = { id: "1" };
        const { container } = render(await PostPageWrapper({ params }));

        const title = await screen.findByText("Sample Post");
        const body = await screen.findByText("This is a sample post body.");
        expect(title).toBeInTheDocument();
        expect(body).toBeInTheDocument();

        expect(global.fetch).toHaveBeenCalledWith(
            `https://jsonplaceholder.typicode.com/posts/1`
        );
        expect(container).toMatchSnapshot(); // Optional: Snapshot testing for the rendered component
    });

    test("throws an error when API call fails", async () => {
        (global.fetch as jest.Mock).mockResolvedValueOnce({
            ok: false,
        });

        const params = { id: "1" };

        await expect(PostPageWrapper({ params })).rejects.toThrow("Failed to fetch post");

        expect(global.fetch).toHaveBeenCalledWith(
            `https://jsonplaceholder.typicode.com/posts/1`
        );
    });

    test("throws an error if fetch response is malformed", async () => {
        (global.fetch as jest.Mock).mockResolvedValueOnce({
            ok: true,
            json: async () => ({}), // Empty object to simulate malformed response
        });

        const params = { id: "1" };

        await expect(PostPageWrapper({ params })).rejects.toThrow("Malformed response");

        expect(global.fetch).toHaveBeenCalledWith(
            `https://jsonplaceholder.typicode.com/posts/1`
        );
    });

});
