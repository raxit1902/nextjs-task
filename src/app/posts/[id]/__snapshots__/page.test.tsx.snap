// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostPageWrapper Component renders post details correctly when API call is successful 1`] = `
<div>
  <div>
    <div
      class="btn-container"
    >
      <div
        class="btn-left-div"
      >
        <div
          data-tooltip-id="home-icon"
        >
          <a
            class=""
            data-testid="go-back"
            href="/"
          >
            <span />
          </a>
        </div>
      </div>
      <div
        class="btn-right-div"
      >
        <div
          data-tooltip-id="revalidate-icon"
        >
          <button
            class="revalidate-btn "
          >
            <span />
          </button>
        </div>
        <div
          data-tooltip-id="edit-icon"
        >
          <a
            class=""
            data-testid="edit-btn"
            href="/posts/edit/1"
          >
            <span />
          </a>
        </div>
      </div>
    </div>
    <div
      class="post"
    >
      <h1>
        Sample Post
      </h1>
      <p>
        This is a sample post body.
      </p>
    </div>
  </div>
</div>
`;
